src/bit_machine/mod.rs

```diff
@@ -366,66 +367,79 @@ impl BitMachine {
     }
 
     fn exec_jet<J: Jet>(&mut self, jet: J, env: &J::Environment) -> Result<(), JetFailed> {
-        use simplicity_sys::c_jets::frame_ffi::{c_readBit, c_writeBit, CFrameItem};
-        use simplicity_sys::c_jets::uword_width;
-        use simplicity_sys::ffi::UWORD;
+        use crate::ffi::c_jets::frame_ffi::{c_readBit, c_writeBit, CFrameItem};
+        use crate::ffi::c_jets::uword_width;
+        use crate::ffi::ffi::UWORD;
+
+        /// Return a read frame that is equivalent to the active read frame.
+        unsafe fn get_input_frame(
+            mac: &mut BitMachine,
+            bit_width: usize,
+            buffer: &mut Vec<UWORD>,
+        ) -> CFrameItem {
+            let buffer_width = uword_width(bit_width);
+            buffer.clear();
+            buffer.resize(buffer_width, 0);
+
+            // Write bits from active read frame into input frame
+            let buffer_end = buffer.as_mut_ptr().add(buffer_width);
+            let mut write_frame = CFrameItem::new_write(bit_width, buffer_end);
+            for _ in 0..bit_width {
+                let bit = mac.read_bit();
+                c_writeBit(&mut write_frame, bit);
+            }
+            mac.back(bit_width);
 
-        // Sanity Check: This should never really fail, but still good to do
-        if !simplicity_sys::c_jets::sanity_checks() {
-            return Err(JetFailed);
+            // Return input frame as read frame
+            let buffer_ptr = buffer.as_mut_ptr();
+            CFrameItem::new_read(bit_width, buffer_ptr)
         }
-        let src_ty_bit_width = jet.source_ty().to_bit_width();
-        let target_ty_bit_width = jet.target_ty().to_bit_width();
-
-        let a_frame_size = uword_width(src_ty_bit_width);
-        let b_frame_size = uword_width(target_ty_bit_width);
-        // a_frame_size + b_frame_size must be non-zero unless it is a unit to unit jet
-        if a_frame_size == 0 && b_frame_size == 0 {
-            return Ok(());
+
+        /// Return a write frame that is equivalent to the active write frame.
```

Comment:
In 494a232fa8e6144be845940a1bfe22c88e98359e:

This comment should just say "returns a new output frame backed by `buffer` of width `bit_width`". And have similar warnings about the safety of `buffer` and maybe say that `bit_width` is expected to be the size of the active write frame.